maltoolbox/model.py

```diff
@@ -53,28 +58,27 @@ def add_asset(
         Return:
         An asset matching the name if it exists in the model.
         """
-        if asset_id is not None:
-            for existing_asset in self.assets:
-                if asset_id == existing_asset.id:
-                    raise ValueError(f'Asset index {asset_id} already in use.')
-            asset.id = asset_id
-        else:
-            asset.id = self.next_id
+
+        # Set asset ID and check for duplicates
+        asset.id = asset_id or self.next_id
+        if asset.id in self.asset_ids:
+            raise ValueError(f'Asset index {asset_id} already in use.')
+        self.asset_ids.add(asset.id)
+
         self.next_id = max(asset.id + 1, self.next_id)
 
         asset.associations = []
 
         if not hasattr(asset, 'name'):
             asset.name = asset.metaconcept + ':' + str(asset.id)
         else:
-            for ex_asset in self.assets:
-                if ex_asset.name == asset.name:
-                    if allow_duplicate_names:
-                        asset.name = asset.name + ':' + str(asset.id)
-                        break
-                    else:
-                        raise ValueError(f'Asset name {asset.name} is a '
-                        'duplicate and we do not allow duplicates.')
+            if asset.name in self.asset_names:
+                if allow_duplicate_names:
+                    asset.name = asset.name + ':' + str(asset.id)
```

Comment:
Irrelevant but the docstring says the name will be prefixed with the id, not suffixed.

Reply:
Thanks!